src/librbd/migration/NBDStream.cc

```diff
@@ -0,0 +1,263 @@
+struct NBDStream<I>::ReadRequest {
+  NBDStream* nbd_stream;
+  io::Extents byte_extents;
+  bufferlist* data;
+  Context* on_finish;
+
+  ReadRequest(NBDStream* nbd_stream, io::Extents&& byte_extents,
+              bufferlist* data, Context* on_finish)
+    : nbd_stream(nbd_stream), byte_extents(std::move(byte_extents)),
+      data(data), on_finish(on_finish) {
+  }
+
+  void send() {
+    data->clear();
+    read();
+  }
+
+  void read() {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << "byte_extents=" << byte_extents << dendl;
+
+    for (auto [byte_offset, byte_length] : byte_extents) {
+      ldout(cct, 20) << "byte_offset=" << byte_offset << " byte_length="
+                     << byte_length << dendl;
+      auto ptr = buffer::ptr_node::create(buffer::create_small_page_aligned(
+        byte_length));
+      int rc = nbd_pread(nbd_stream->m_nbd, ptr->c_str(), byte_length,
+                         byte_offset, 0);
+      data->push_back(std::move(ptr));
+      if (rc == -1) {
+        rc = nbd_get_errno();
+        lderr(cct) << "nbd_pread: " << nbd_get_error()
+                   << " (errno=" << rc << ")" 
+                   <<  byte_offset << " " << byte_length << dendl;
+        on_finish->complete(rc);
+        return;
+      }
+    }
+
+    finish(0);
+  }
+
+  void finish(int r) {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << "r=" << r << dendl;
+
+    if (r < 0) {
+      data->clear();
+    }
+
+    on_finish->complete(r);
+    delete this;
+  }
+};
+
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::migration::NBDStream::ListSparseExtentsRequest: " \
+                           << this << " " << __func__ << ": "
+
+template <typename I>
+struct NBDStream<I>::ListSparseExtentsRequest {
+  NBDStream* nbd_stream;
+  io::Extents byte_extents;
+  io::SparseExtents* sparse_extents;
+  Context* on_finish;
+
+  ListSparseExtentsRequest(NBDStream* nbd_stream, io::Extents&& byte_extents,
+                           io::SparseExtents* sparse_extents, Context* on_finish)
+    : nbd_stream(nbd_stream), byte_extents(std::move(byte_extents)),
+      sparse_extents(sparse_extents), on_finish(on_finish) {
+  }
+
+  void send() {
+    list_sparse_extents();
+  }
+
+  void list_sparse_extents() {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << "byte_extents=" << byte_extents << dendl;
+
+    for (auto [byte_offset, byte_length] : byte_extents) {
+      ldout(cct, 20) << "byte_offset=" << byte_offset << " byte_length="
+                     << byte_length << dendl;
+      int rc = nbd_block_status(nbd_stream->m_nbd, byte_length, byte_offset,
+                                {extent_cb, sparse_extents}, 0);
+      if (rc == -1) {
+        rc = nbd_get_errno();
+        lderr(cct) << "nbd_block_status: " << nbd_get_error()
+                   << " (errno=" << rc << ")" 
+                   << byte_offset << " " << byte_length << dendl;
+        on_finish->complete(rc);
+        return;
```

Comment:
You were probably commenting on individual commits instead of the final diff.  These lines aren't there in the final diff:
```
        // don't propagate errors -- we are set up to list any missing
        // parts of the range as DATA if nbd_block_status() returns less
        // status or none at all
```